### ISSUE TITLE
Alert about "nearby speed camera" devices, even though they are not "properly mapped"

### ISSUE THREAD

**Issue by Max1234-Ita** [osmandapp/OsmAnd#19501]
### Describe the idea (required)

I'm using OsmAnd since a long-time (12 years or so), and over this time I could interact with many users by providing some help or exchanging opinions.

One of the complaints I've heard most frequently is that OsmAnd "does not warn about speed cameras". 

It does, of course, but not for all the devices present in the map; Apart from the lack of mapping, the other reason for speed cameras not being announced is that OsmAnd alerts about them only if they are "properly mapped" in OSM.

In OpenStreetMap, a Speed Camera is "properly mapped" if:
- A relation of type _enforcement_ exists, and the enforcement type is specified (i.e. _'maxspeed'_, _'average_speed'_, _'traffic_signals'_)
- The relation contains one way having _'from'_  role;
- The relation contains one way having _'to'_ 
- The relation contains a node with _'device'_ role; the node is usually (always, I guess) tagged as highway=speed_camera.

Unfortunately, there are in the map a lot of _highway=speed_camera_ elements that do not belong to any enforcement relation: I guess that the main reason is because they have been added by inexperienced users or through mapping apps that do not fully support the creation of relations.

I made some query through the Overpass Wizard (https://overpass-turbo.eu):
1. _highway=speed_camera in Italy_ --> Got a GeoJSON file containing 4342 _"highway": "speed_camera"_ elements;
2. _type=enforcement and (enforcement=maxspeed or enforcement=average_speed or enforcement=traffic_signals) in Italy_ --> The resulting GeoJSON contained 1929 elements marked with  _"role": "device"_ 

By doing some maths: 4342 -  1929 = 2413  (!) 
So, 2413 speed cameras (in Italy) are not properly mapped nor reported by OsmAnd (that's about 55.6%); I tried the same queries for France and the situation was quite better, having about 17% of SC not belonging to any relation (it's 800 elements over 4560, anyway)

So here's my proposal: why not improve speed camera detection by raising an alert also if a _highway=speed_camera_ element is found within a certain range?

This would have 2 benefits:
1. Extend the number of object recognized and announced by OsmAnd, improving the app usability.
2. If the voice prompt and/or the icon used is different from the "standard" one, this might help experienced mappers recognizing the mapping problem and possibly have them to correct those issues.

As a "contra", speed camera announcements made like that might be a bit unreliable, due to the lesser precision of the "nearby" metod vs. the already implemented one (which is good and should remain in use, of course); this would not be worse than the behaviour of other navigation apps, however.

### Tell us about the expected behaviour (required)

- In addition to the usual speed camera alerts, a new Display/Voice one, i.e. "Nearby Speed Camera", should be issued if a _highway=speed_camera_ element is found within a certain range/time from current position (i.e. 200 meters or 10 seconds)

- In order to limit the number of false-positive cases, the issue should be raised if the approaching speed is similar to the current speed (i.e. at least 70% of it): this should avoid that, i.e. an announcement for traffic signals or urban speed limit is raised while driving along a motorway;

- The new alert should NOT be raised if OsmAnd is in Navigation Mode (a route has been calculated AND the element belongs to an "enforcement" relation (-> already recognized by OsmAnd, no need to announce twice);

- The new alert might be raised if OsmAnd is not nevigating, even if the element belongs to an "enforcement" relation (not sure if by current behaviour SCs are announce while not navigating);

- The new voice prompt might be turned on/off by User preference in _Configure profile -> Voice prompts_ menu, either linking it to the existing "Speed cameras" slider or adding a dedicated "Nearby speed cameras" option (best solution);

- The new screen alert might be turned on/off by User preference in _Configure profile ->
Screen alerts, either linking it to the existing "Speed cameras" slider or adding a dedicated "Nearby speed cameras" option (best solution)

### Tell us about alternatives you've considered (required)

Currently, as a mapper, the only alternative is to hear some user complaining about a speed camera not being announced, go for a check in the map, find that it's not properly mapped (80% of the cases) and add the required enforcement relation. 

Unfortunately, implementing a proper mapping requires some experience that not all users have, so in many cases the situation is not being corrected.

### Context (optional)

Any comment is welcome!
(maybe some user will add the Overpass query results for his/her Country)

**Comment by mariush444** [comment]
Speedcam has to attachet to road additionally AFAIK.

**Comment by Max1234-Ita** [comment]
> Speedcam has to attachet to road additionally AFAIK.

Nope... unfortunately, in OSM the *enforcement* is ON the road, the *device* is a floating nod besides the road... in most cases. Anyway it's not necesssarily on the road (and inexperienced users just add the device node and forget the relation).

**Comment by vshcherb** [comment]
We plan to implement that feature however  as it addressed it's not related to OpenStreetMap data cause there enforcement should indicate start/end point on the road and device is a separate node.

**Comment by Max1234-Ita** [comment]
> We plan to implement that feature however as it addressed it's not related to OpenStreetMap data cause there enforcement should indicate start/end point on the road and device is a separate node.

Yes, this feature is not strictly related to OpenStreet mapping, it's more about bad mapping from inexperienced people or users of "automated mapping" apps (i.e. StreetComplete), which only allow to add the device node but do not create the full _enforcement_ relation.

However, the feature would help a lot those people who use OsmAnd as a pure car/motorbike navigator: knowing that a speedcamera is approaching would be very useful, even though the device it's not perfectly mapped.

**Comment by pebogufi** [comment]
For that purpose i use this workaround since years:
- Create a custom search with the tags you want to be informed.
- before starting a navigation connect this poi-overlay to your routing profile.
- in this profile in navigation setting / voice prompt / "Nearby POI" must be activated

When you have prepared navigation you can set/change the alert distance
Open navigation setting/ show along route / point of interest.
Now you can change the value "within".
Have fun 😉

**Comment by gr4v1ty69** [comment]
+1 for this idea as there's no viable alternative for France and alerts for Speed Cameras.

**Comment by elgaard** [comment]
We should just fix the bad mapping and educate new users.
I have used overpass turbo to find speed cameras that are not on roads or in enforcement relations:

https://overpass-turbo.eu/s/23iT

There are about 15000 of those.
In addition there are some enforcement relations missing a "to" node on the road (I have not yes figures out how to search for that)
I have already fixed most speed camera tagging in France and Denmark.
I have inspected many of the places using photo overlays. Quite a few "speed cameras" turn out to instead be speed displays.

Recent bad speed camera tagging:
https://overpass-turbo.eu/s/23iS

I have left comments on many of the changesets a couple of days ago and already gotten very good responses.
Two users even said that after reading the wiki, they realized that the speed cameras were instead speed displays.

Also it does not help that OsmAnd creates speed cameras that it itself cannot use because the tagging is not correct. I will create an issue about that.

**Comment by vmicho** [comment]
I'd personally prefer if osmands add some kind of warning, can be in voice, like "Incorrectly mapped speedcam nearby"

**Comment by Danysan1** [comment]
Regardless of how OsmAnd will handle these incorrect speed cameras, the appropriate places to add this kind of checks would be the [Osmose](https://osmose.openstreetmap.fr/en/map/) QA tool and the [iD editor validation presets](https://github.com/openstreetmap/iD/tree/develop/modules%2Fpresets), so that more mappers will be aware of the existing errors and hopefully stop creating new ones

**Comment by Max1234-Ita** [comment]
> Regardless of how OsmAnd will handle these incorrect speed cameras, the appropriate places to add this kind of checks would be the [Osmose](https://osmose.openstreetmap.fr/en/map/) QA tool and the [iD editor validation presets](https://github.com/openstreetmap/iD/tree/develop/modules%2Fpresets), so that more mappers will be aware of the existing errors and hopefully stop creating new ones

I agree and would also include the validation checks in JOSM, which is used by a very large number of mappers.

IMO, in order to solve the problem at its root, some check (and mapping procedure) should be applied also in the field mapping apps, such as StreetComplete etc. unfortunately that cannont be discussed here. 

The issues are mainly caused by inexperienced mappers who think that mapping a single node where they see a device is enough to have it reported as a potential threat during their trips.

So either the mappers operate as expected or the apps use less strict methods to detect the enforcements (the problem does not only affect speed checks)... or -better- both the things.

**Comment by elgaard** [comment]
I agree.

I did create a MapRoulette challenge:
https://maproulette.org/browse/challenges/52030

But nothing happened yet.
Maybe you can promote it?

**Comment by Danysan1** [comment]
Turns out osmose already has this check: https://osmose.openstreetmap.fr/en/map/#issue_uuid=dd83f727-6a95-d015-1149-54be991985b5&item=1280&level=1%2C2%2C3
